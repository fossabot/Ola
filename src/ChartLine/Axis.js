import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const ChartLineAxis = ({ max, steps, className }) => {
  const styles = cx('ola_chartLine-axis', className)

  steps = Math.min(max, steps)

  const step = max / steps
  const children = new Array(steps + 1).fill(null)
    .map((v, index) => <span key={index}>{Math.round(index * step * 10) / 10}</span>)
  
  return (
    <div className={styles}>
      { children }
    </div>
  )
}

ChartLineAxis.defaultProps = {
  steps: 4,
  max: null,
  className: null
}

ChartLineAxis.propTypes = {
  /** Max value */
  max: PT.number.isRequired,
  /** Number of steps */
  steps: PT.number,
  /** Extra className */
  className: PT.string,
}

export default ChartLineAxis
