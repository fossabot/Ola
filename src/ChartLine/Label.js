import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const Label = ({ value, children, className }) => {
  const styles = cx('ola_chartLine-label', className)
  return (
    <div className={styles} style={{ '--value': value }}>
      { children }
    </div>
  )
}

Label.defaultProps = {
  value: 0,
  className: null
}

Label.propTypes = {
  /** Value between 0 and 1 */
  value: PT.number,
  /** Extra className */
  className: PT.string,
  /** Childen nodes */
  children: PT.oneOfType([
    PT.arrayOf(PT.node),
    PT.node
  ])
}

export default Label