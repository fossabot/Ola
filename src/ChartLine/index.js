import React from 'react'
import {default as PT} from 'prop-types'

const ChartLine = ({ children, values }) => {
  const draw = drawPath(values)
  const lineD = `M 0 0 ${draw}`
  const fillD = `M 0 0 ${draw} V 0 H 0 Z`

  return (
    <div className="ola_chartLine">
      { children }

      <svg role="img" viewBox="0 0 100 160" xmlns="http://www.w3.org/2000/svg" preserveAspectRatio="none" className="ola_chartLine-svg">
        <path d={fillD} vectorEffect="non-scaling-stroke" strokeLinejoin="round" className="ola_chartLine-svg-background" />
        <path d={lineD} vectorEffect="non-scaling-stroke" strokeLinejoin="round"  className="ola_chartLine-svg-line" />
      </svg>
    </div>
  )
}

ChartLine.defaultProps = {
  values: [],
}

ChartLine.propTypes = {
  /** All values */
  values: PT.arrayOf(PT.number).isRequired,
  /** Childen nodes */
  children: PT.oneOfType([
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired
}

export default ChartLine

function drawPath(values) {
  const distance = 100 / values.length
  const commands = [`M 0 ${getYPosition(values[0])}`]
  let x = distance / 2

  values.forEach((value) => {
    commands.push(`L ${x} ${getYPosition(value)}`)
    x += distance
  })

  commands.push(`L 100 ${getYPosition(values[values.length - 1])}`)

  return commands.join(' ')
}

function getYPosition(value) {
  return 150 - (value * 150) + 5
}