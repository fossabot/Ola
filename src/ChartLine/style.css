.ola_chartLine {
    --height: 150px;
    --color: var(--brand);

    display: flex;
    position: relative;
    text-align: center;
    z-index: 1;
    min-height: calc(var(--height) + 10px);

    &.ola-skeleton {
        --color: var(--gray-xlight);

        & .ola_chartLine-label::before {
            visibility: hidden;
        }
        &.is-loading .ola_chartLine-svg {
            background-size: 800px 100%;
            animation: loadingSkeleton 2s linear infinite;
        }
    }

    &.is-multiple {
        & .ola_chartLine-svg-background {
            fill: none;
        }
    }
}

.ola_chartLine-svg {
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    height: calc(var(--height) + 10px);
    z-index: -1;
}
.ola_chartLine-svg-background {
    fill: var(--white);
    stroke: none;
}
.ola_chartLine-svg-line {
    fill: none;
    stroke: var(--color);
    stroke-width: 4px;
}

.ola_chartLine-value {
    position: relative;
    flex-grow: 1;
    flex-basis: 10px;
    font: var(--font-callout);
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    border-bottom: solid 1px var(--gray-xlight);
}
.ola_chartLine-space {
    height: calc((var(--height) * var(--value)) + var(--size-6));
    flex-shrink: 0;
}
.ola_chartLine-point {
    border: solid 4px var(--color);
    background: white;
    width: 8px;
    height: 8px;
    border-radius: 50%;
    margin: auto;
    display: block;
    position: absolute;
    left: 50%;
    margin-left: -8px;
    bottom: calc((var(--height) * var(--value)) - 4px);
}
.ola_chartLine-footer {
    height: 0;
    position: relative;
    top: var(--size-3);
}
.ola_chartLine-rule {
    position: absolute;
    left: 0;
    width: 100%;
    border-bottom: dashed 1px var(--gray-light);
    padding-bottom: var(--size-2);
    bottom: calc(var(--value) * (var(--height) + 5px));
    text-align: right;
    mix-blend-mode: multiply;
    font: var(--font-caption);
    color: var(--gray);
    pointer-events: none;
}
.ola_chartLine-axis {
    position: relative;
    display: none;
    flex-direction: column-reverse;
    z-index: 1;
    align-self: flex-end;
    height: calc(var(--height) + var(--size-4));
    background: var(--white);
    justify-content: space-between;

    border-right: solid 1px var(--gray-light);
    text-align: right;
    font: var(--font-caption);
    color: var(--gray);
    padding-right: var(--size-3);
    width: var(--size-8);
    white-space: nowrap;
    box-sizing: border-box;
    flex-shrink: 0;


    @media (min-width: 500px) {
        display: flex;

        & ~ .ola_chartLine-svg {
            left: var(--size-8);
            width: calc(100% - var(--size-8));
        }
    }
}