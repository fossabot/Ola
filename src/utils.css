/* Deprecated, use .ola-bg-gray-xlight instead */
.ola-bg {
    background-color: var(--gray-xlight);
}

/* Text sizes */
.ola-display {
    font: var(--font-display);
    letter-spacing: var(--font-display-spacing);
}
.ola-title {
    font: var(--font-title);
    letter-spacing: var(--font-title-spacing);
}
.ola-headline {
    font: var(--font-headline);
}
.ola-body {
    font: var(--font-body);
}
.ola-callout {
    font: var(--font-callout);
}
.ola-caption {
    font: var(--font-caption);
}

/* Colors */
.ola-brand {
    color: var(--brand);
}
.ola-white {
    color: var(--white);
}
.ola-black {
    color: var(--black);
}
.ola-gray {
    color: var(--gray);
}
.ola-gray-light {
    color: var(--gray-light);
}
.ola-gray-xlight {
    color: var(--gray-xlight);
}
.ola-error {
    color: var(--error);
}
.ola-warning {
    color: var(--warning);
}
.ola-success {
    color: var(--success);
}
.ola-pro {
    color: var(--pro);
}

/* Background colors */
.ola-bg-brand {
    background-color: var(--brand);
}
.ola-bg-white {
    background-color: var(--white);
}
.ola-bg-black {
    background-color: var(--black);
}
.ola-bg-gray {
    background-color: var(--gray);
}
.ola-bg-gray-light {
    background-color: var(--gray-light);
}
.ola-bg-gray-xlight {
    background-color: var(--gray-xlight);
}
.ola-bg-error {
    background-color: var(--error);
}
.ola-bg-warning {
    background-color: var(--warning);
}
.ola-bg-success {
    background-color: var(--success);
}
.ola-bg-pro {
    background-color: var(--pro);
}

/* Alignment */
.ola-right {
    text-align: right;
}
.ola-center {
    text-align: center;
}
.ola-numeric {
    font-variant-numeric: tabular-nums;
    text-align: right;
}

/* Misc */
.ola-nomargin {
    margin: 0;
}
.ola-bold {
    font-weight: var(--bold);
}