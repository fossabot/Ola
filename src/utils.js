/**
 * getElementType
 *
 * Returns a createElement() type based on the props of the Component.
 * Useful for calculating what type a component should render as.
 * @param {function} Component A function or ReactClass.
 * @param {object} props A ReactElement props object
 * @returns {string|function} A ReactElement type
*/

const getElementType = (Component, props) => {
  const { defaultProps = {} } = Component
  if (props.as && props.as !== defaultProps.as) return props.as
  if (props.href) return 'a'
  return defaultProps.as || 'div'
}

export { getElementType }




/**
 * normalizeRange
 *
 * Returns a number with a normalizade range based on the max and min value.
 * Useful for calculating ChartColumnValue values where is a number between 0 and 1.
 * @param {val} Value of current element
 * @param {max} Max value of all elements 
 * @param {min} Min value of all elements 
 * @returns {number} 
*/
const normalizeRange = (val, max, min = 0) => (val - min) / (max - min)
export { normalizeRange }

export function getRanges(values) {
  const max = Math.max(...values)
  return values.map(value => normalizeRange(value, max))
}