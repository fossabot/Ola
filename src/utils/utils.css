/* Deprecated, use .ola-bg-gray-xlight instead */
.ola-bg {
    background-color: var(--gray-xlight);
}

/* Text sizes */
.ola-display {
    font: var(--font-display);
    letter-spacing: var(--font-display-spacing);
}
.ola-title {
    font: var(--font-title);
    letter-spacing: var(--font-title-spacing);
}
.ola-headline {
    font: var(--font-headline);
}
.ola-body {
    font: var(--font-body);
}
.ola-callout {
    font: var(--font-callout);
}
.ola-caption {
    font: var(--font-caption);
}

h1,
h2,
h3,
h4,
h5,
h6,
b,
strong {
    &.ola-headline,
    &.ola-body,
    &.ola-callout,
    &.ola-caption {
        font-weight: var(--bold);
    }
}

/* Colors */
.ola-brand {
    color: var(--brand);
}
.ola-accent {
    color: var(--accent);
}
.ola-white {
    color: var(--white);
}
.ola-black {
    color: var(--black);
}
.ola-gray {
    color: var(--gray);
}
.ola-gray-light {
    color: var(--gray-light);
}
.ola-gray-xlight {
    color: var(--gray-xlight);
}
.ola-error {
    color: var(--error);
}
.ola-warning {
    color: var(--warning);
}
.ola-success {
    color: var(--success);
}
.ola-pro {
    color: var(--pro);
}
.ola-black-bolds b,
.ola-black-bolds strong {
    color: var(--black);
}

/* Background colors */
.ola-bg-brand {
    background-color: var(--brand);
}
.ola-bg-accent {
    background-color: var(--accent);
}
.ola-bg-white {
    background-color: var(--white);
}
.ola-bg-black {
    background-color: var(--black);
}
.ola-bg-gray {
    background-color: var(--gray);
}
.ola-bg-gray-light {
    background-color: var(--gray-light);
}
.ola-bg-gray-xlight {
    background-color: var(--gray-xlight);
}
.ola-bg-error {
    background-color: var(--error);
}
.ola-bg-warning {
    background-color: var(--warning);
}
.ola-bg-success {
    background-color: var(--success);
}
.ola-bg-pro {
    background-color: var(--pro);
}

/* Alignment */
.ola-right {
    text-align: right;
}
.ola-center {
    text-align: center;
}
.ola-numeric {
    font-variant-numeric: tabular-nums;
    text-align: right;
}

/* Misc */
.ola-nomargin {
    margin: 0;
}
.ola-nomargin-top {
    margin-top: 0;
}
.ola-nomargin-bottom {
    margin-bottom: 0;
}
.ola-bold {
    font-weight: var(--bold);
}