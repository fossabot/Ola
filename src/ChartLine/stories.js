import React from 'react'
import ChartLine from './'
import ChartLineLabel from './Label'
import { getRanges } from '../utils'

const mockData = [
  {
    value: 50,
    label: '50 keywords',
  },
  {
    value: 20,
    label: '20 keywords',
  },
  {
    value: 0,
    label: '0 keywords',
  },
  {
    value: 100,
    label: '100 keywords',
  }
]

const mockData2 = [
  {
    value: 23,
    label: '23 sites',
  },
  {
    value: 10,
    label: '10 sites',
  },
  {
    value: 0,
    label: '0 sites',
  },
  {
    value: 90,
    label: '90 sites',
  }
]

const colors = [
  'var(--brand)',
  'var(--error)',
  'pink',
  'var(--success)',
]

export default {
  title: 'ChartLine',
  component: ChartLine,
  args: {}
}

export const Base = () => {
  const ranges = getRanges(mockData.map( data => data.value ))
  return (
    <ChartLine ranges={ranges} colors={colors}>
      { mockData.map((row, idx) => 
        <ChartLineLabel key={idx} value={ranges[idx]}>
          {row.label}
        </ChartLineLabel>
      )}
    </ChartLine>
  )
}

export const Multiple = () => {
  const ranges = getRanges(mockData.map( data => data.value ), mockData2.map( data => data.value ))
  return (
    <ChartLine ranges={ranges} colors={colors}>
      { mockData.map((row, idx) => 
        <ChartLineLabel key={idx} value={ranges[idx]}>
          {row.label} <br/> from {mockData2[idx].label}
        </ChartLineLabel>
      )}
    </ChartLine>
  )
}

export const HtmlLabels = () => {
  const ranges = getRanges(mockData.map( data => data.value ))

  return (
    <ChartLine ranges={ranges}>
      { mockData.map((row, idx) => 
        <ChartLineLabel key={idx} value={ranges[idx]}>
          <div><strong>{row.label}</strong><br />{row.label}</div>
        </ChartLineLabel>
      )}
    </ChartLine>
  )
}
