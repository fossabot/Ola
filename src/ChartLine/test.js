import React from 'react'
import ChartLine from './'
import ChartLineLabel from './Label'
import { getRanges } from '../utils'
import renderer from 'react-test-renderer'

const mockData = [
  {
    value: 5,
    label: '5 keywords'
  },
  {
    value: 2,
    label: '2 keywords'
  },
  {
    value: 1,
    label: '1 keywords'
  },
  {
    value: 10,
    label: '10 keywords'
  }
]

it('Default ChartLine', () => {
  const tree = renderer
    .create(<ChartLine>Test</ChartLine>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Default ChartLineLabel', () => {
  const tree = renderer
    .create(<ChartLineLabel></ChartLineLabel>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('ChartLine and ChartLineLabel', () => {
  const ranges = getRanges(mockData.map(data => data.value))
  const tree = renderer
    .create(
      <ChartLine>
        { mockData.map((row, idx) => 
          <ChartLineLabel key={idx} value={ranges[idx]}>{row.label}</ChartLineLabel>
        )}
      </ChartLine>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})


