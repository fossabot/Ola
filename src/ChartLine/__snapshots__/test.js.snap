// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLine and ChartLineLabel 1`] = `
<div
  className="ola_chartLine"
>
  <div
    className="ola_chartLine-label"
    style={
      Object {
        "--value": 0.5,
      }
    }
  >
    5 keywords
  </div>
  <div
    className="ola_chartLine-label"
    style={
      Object {
        "--value": 0.2,
      }
    }
  >
    2 keywords
  </div>
  <div
    className="ola_chartLine-label"
    style={
      Object {
        "--value": 0.1,
      }
    }
  >
    1 keywords
  </div>
  <div
    className="ola_chartLine-label"
    style={
      Object {
        "--value": 1,
      }
    }
  >
    10 keywords
  </div>
  <svg
    className="ola_chartLine-svg"
    preserveAspectRatio="none"
    role="img"
    viewBox="0 0 100 160"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="ola_chartLine-svg-background"
      d="M 0 0 M 0 80 L 12.5 80 L 37.5 125 L 62.5 140 L 87.5 5 L 100 5 V 0 H 0 Z"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
    <path
      className="ola_chartLine-svg-line"
      d="M 0 0 M 0 80 L 12.5 80 L 37.5 125 L 62.5 140 L 87.5 5 L 100 5"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
  </svg>
</div>
`;

exports[`Default ChartLine 1`] = `
<div
  className="ola_chartLine"
>
  Test
  <svg
    className="ola_chartLine-svg"
    preserveAspectRatio="none"
    role="img"
    viewBox="0 0 100 160"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      className="ola_chartLine-svg-background"
      d="M 0 0 M 0 0 L 100 0 V 0 H 0 Z"
      strokeLinejoin="round"
      vectorEffect="non-scaling-stroke"
    />
  </svg>
</div>
`;

exports[`Default ChartLineLabel 1`] = `
<div
  className="ola_chartLine-label"
  style={
    Object {
      "--value": 0,
    }
  }
/>
`;
